### ISSUE TITLE
How do you perform clustering after you get a binary segmentation?

### ISSUE THREAD

**Issue by yinhai86924** [wvangansbeke/LaneDetection_End2End#13]
Is the final loss function derived from the lane geometry model?  as shown in the paper？

**Comment by wvangansbeke** [comment]
Hi @yinghai86924,

The output of the network contains 4 channels since we want to detect four lane lines. Otherwise you would not be able to distinguish the different lane lines from eachother. So we don't do binary semantic segmentation since there are more than 2 output classes for every pixel.

These outputs are the weight maps for the least squares layer. In order to know which weight map corresponds to which lane line, we fix the order. The order is always relative to car. In this way, the loss function can become a simple regression loss. However, we found that the area loss as explained in the paper was more robust. This means that we don't do any clustering. 

Hope this helps, 
Wouter

**Comment by jiayily** [comment]
Hi @wvangansbeke ,
Why do you remove the background? Usually the output of the segmentation network should contain 5 channels when we want to detect four lane lines. And a softmax layer is used to pick out the most possible class. Your practice is a little strange to me.
